### PR TITLE
Http::Range integration for file/datalake

### PR DESCRIPTION
**Pull request by katmsft**
Resolves https://github.com/Azure/azure-sdk-for-cpp/issues/1245

# Pull Request Checklist

Please leverage this checklist as a reminder to address commonly occurring feedback when submitting a pull request to make sure your PR can be reviewed quickly:

See the detailed list in the [contributing guide](https://github.com/Azure/azure-sdk-for-cpp/blob/master/CONTRIBUTING.md#pull-requests).

- [x] [C++ Guidelines](https://azure.github.io/azure-sdk/cpp_introduction.html)
- [x] Doxygen docs
- [x] Unit tests
- [x] No unwanted commits/changes
- [x] Descriptive title/description
  - [x] PR is single purpose
  - [x] Related issue listed
- [x] Comments in source
- [x] No typos
- [x] Update changelog
- [x] Not work-in-progress
- [x] External references or docs updated
- [x] Self review of PR done
- [x] Any breaking changes?

### CODE REVIEW

File: sdk/storage/azure-storage-files-datalake/inc/azure/storage/files/datalake/datalake_options.hpp [hunk 1]
```diff
@@ -731,45 +726,6 @@ namespace Azure { namespace Storage { namespace Files { namespace DataLake {
     Azure::Core::Nullable<std::string> ExpiresOn;
   };
 
-  /**
-   * @brief Optional parameters for FileClient::DownloadToBuffer and FileClient::DownloadToFile.
-   */
-  struct DownloadFileToBufferOptions
```
Comment (Jinming-Hu) [inline sdk/storage/azure-storage-files-datalake/inc/azure/storage/files/datalake/datalake_options.hpp:737]: why was this deleted?

**Review by Jinming-Hu — APPROVED** [review]

File: sdk/storage/azure-storage-files-shares/src/share_file_client.cpp [hunk 2]
```diff
@@ -202,18 +202,18 @@ namespace Azure { namespace Storage { namespace Files { namespace Shares {
       const DownloadFileOptions& options) const
   {
     auto protocolLayerOptions = Details::ShareRestClient::File::DownloadOptions();
-    if (options.Offset.HasValue())
```
Comment (Jinming-Hu) [inline sdk/storage/azure-storage-files-shares/src/share_file_client.cpp:205]: You copied these changes from blob service, right? I didn't review this file since the logic here is complicated

File: sdk/storage/azure-storage-files-datalake/inc/azure/storage/files/datalake/datalake_options.hpp (hunk 1 above)
Comment (katmsft) [reply sdk/storage/azure-storage-files-datalake/inc/azure/storage/files/datalake/datalake_options.hpp:737]: We don't have any API that's using these options anymore.
